---
title: 'TBET expression and activity in tonsillar naive and memory B-cells'
author: "Ramon Massoni-Badosa"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, out.width = "100%", fig.align='center', 
                      message=FALSE, warning = FALSE)
options(width = 1200)
```


# Introduction

The aim of this notebook is to quantify the expression and activity of [TBX21](https://www.genecards.org/cgi-bin/carddisp.pl?gene=TBX21) (TBET) in naive and memory B-cells in human tonsil. To this end, we will use our [tonsil cell atlas](https://www.biorxiv.org/content/10.1101/2022.06.24.497299v1.full.pdf), which is a single-cell driven taxonomy of cell types and states in a human tonsil in the context of [the Human Cell Atlas](https://elifesciences.org/articles/27041).


TBET+ B-cells are also known as age-associated B-cells (ABC), as  their frequency increases as a function of age, stabilizing at the age of 30 in heatlhy humans. ABC accumulate in bacterial, viral and parasitic infections. In addition, their frequency also increases in autoimmune diseases such as scleroderma or rheumatoid arthritis. Recently, [TBET expression has been linked to FCRL5+ memory B-cells](https://www.cell.com/immunity/fulltext/S1074-7613(23)00094-8?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS1074761323000948%3Fshowall%3Dtrue), so we hypothesize that TBET will be more expressed and active in the FCRL5+ MBC cells of the tonsil atlas.


# Pre-processing

## Load packages

```{r}
library(Seurat)
library(readr)
library(tidyr)
library(stringr)
library(dplyr)
library(ggplot2)
library(purrr)
library(tibble)
library(forcats)
library(Nebulosa)
library(ggrastr)
library(ggpmisc)
library(readxl)
library(scCustomize)
library(UCell)
library(here)
```


## Define parameters and variables

```{r}
cols_annotation <- c(
  "NBC" = "#dbeef2",
  "NBC early activation" = "#95cce0",
  "NBC IFN-activated" = "#4d828c",
  "NBC CD229+" = "#8a73ae",
  "Early GC-commited NBC" = "#74c9b8",
  "GC-commited NBC" = "#6dcbb4",
  "preGC" = "#6ebf90",
  "Proliferative NBC" = "#205b69",
  "GC DZ Noproli" = "#558e1b",
  "Early MBC" = "#ebe2c3",
  "ncsMBC" = "#dccb9b",
  "ncsMBC FCRL4/5+" = "#bca041",
  "csMBC" = "#6f663a",
  "csMBC FCRL4/5+" = "#a29158",
  "MBC FCRL5+" = "#332d1b"
)
theme_pseudobulk <- function(p) {
  p +
    theme_classic() +
    theme(
      legend.position = "none",
      axis.title.y = element_blank(),
      axis.text = element_text(size = 7),
      axis.title = element_text(size = 8)
    )
}
theme_umap <- function(p) {
  p +
    theme(plot.title = element_blank(),
        axis.ticks = element_blank(),
        axis.text = element_text(size = 7),
        axis.title = element_text(size = 8),
        legend.title = element_text(size = 9),
        legend.text = element_text(size = 8))
}
```


## Download and read data

We will download the data from Zenodo (repo: [10.5281/zenodo.6340174](https://zenodo.org/record/6340174)) in the format of a [Seurat object](https://satijalab.org/seurat/articles/essential_commands.html) saved as a [.RDS file](https://www.r-bloggers.com/2016/12/remember-to-use-the-rds-format/). In particular, we download the object "20220215_NBC_MBC_seurat_obj.rds" with the following command:

```{r eval=FALSE}
options(timeout=1000)
dir.create(here("data"))
download.file(
  url = "https://zenodo.org/record/6340174/files/20220215_NBC_MBC_seurat_obj.rds",
  destfile = here("data/20220215_NBC_MBC_seurat_obj.rds")
)
```

We can now load it. Note that the dataset is composed of cells profiled with 3' scRNA-seq and Multiome. To reduce batch effects, we will only work with 3' scRNA-seq data:

```{r}
seurat <- readRDS(here("data/20220215_NBC_MBC_seurat_obj.rds"))
table(seurat$assay)
seurat <- seurat[, seurat$assay == "3P"]
```

In addition, we will load the following files:

- **All_Bcells_level_5_auc.csv**: in the tonsil atlas paper, we ran [pySCENIC](https://www.nature.com/articles/s41596-020-0336-2) to infer the activity of numerous transcription factors (TF) for each tonsillar cell. In the last step of the algorithm, pySCENIC quantfies the activity of each regulon (TF + target genes) into a per-cell score using [AUCell](https://bioconductor.org/packages/release/bioc/html/AUCell.html). Intuitively, AUCell assesses whether the target genes of a TF are enriched amongst the genes with the highest expression in each cell. This file contains, for each cell (rows), the AUCell values for associated with each regulon (columns).
- **All_Bcells_level_5_regulons.csv**: this file is a binary matrix of genes (rows) by regulons (columns); whereby an entry is 1 if the gene is included in the regulon, and 0 otherwise.
- **NBC_MBC_annotation_df.rda**: cell type annotations are intrinsically dynamic, because they can improve as more experts explore the data. This file contains an updated annotation of the naive and memory cell types present in the dataset, which we will add to the Seurat object.
- **Tbx21-dependent genes.xlsx**: excel file that contains the inferred targets of TBET in chronic lymphocytic leukemia (CLL) as part of the present manuscript. These targets are label as "positive" or "negative" depending on whether their expression correlates or anticorrelates with TBET expression.

All these files are available in the "data" folder present in this repository.

```{r}
auc_values <- read_csv(here("data/All_Bcells_level_5_auc.csv"))
regulons <- read_csv(here("data/All_Bcells_level_5_regulons.csv"))
load(here("data/NBC_MBC_annotation_df.rda"))
tbx21_dependent_genes <- read_excel(here("data/Tbx21-dependent genes.xlsx"))
```


## Data wrangling

We will change the name of some variables so they are more significant. We will also subset the AUCell values to keep only TBET activity values, and we will add them to the Seurat object. Finally, we will also add the latest annotation to naive and memory B-cells:

```{r}
colnames(auc_values)[colnames(auc_values) == "...1"] <- "barcode"
colnames(regulons)[colnames(regulons) == "...1"] <- "gene"
colnames(tbx21_dependent_genes)[colnames(tbx21_dependent_genes) == "...1"] <- "gene"
pos_genes <- tbx21_dependent_genes$gene[tbx21_dependent_genes$`Corr. With Tbx21 expression` == "positive"]
neg_genes <- tbx21_dependent_genes$gene[tbx21_dependent_genes$`Corr. With Tbx21 expression` == "negative"]
auc_values <- as.data.frame(auc_values)
rownames(auc_values) <- auc_values$barcode
auc_values <- auc_values[colnames(seurat), ]
if (all(auc_values$barcode == colnames(seurat))) {
  seurat$TBET_activity <- auc_values$`TBX21(+)`
}
NBC_MBC_annotation_df <- NBC_MBC_annotation_df[NBC_MBC_annotation_df$barcode %in% colnames(seurat), ]
rownames(NBC_MBC_annotation_df) <- NBC_MBC_annotation_df$barcode
seurat <- seurat[, NBC_MBC_annotation_df$barcode]
if (all(NBC_MBC_annotation_df$barcode == colnames(seurat))) {
  seurat$annotation_20220619 <- NBC_MBC_annotation_df$annotation_20220619
  Idents(seurat) <- "annotation_20220619"
}
```


# Create pseudobulk profiles of TBET expression and activity

As most TFs, TBET is expressed at low levels. Since lowly expressed genes are frequently not captured in scRNA-seq due to dropout events, we will create pseudobulk expression of TBET for each combination of donor and cell type. Of note, [pseudobulk methods outperform single-cell-tailored methods in differential expression analyses](https://www.nature.com/articles/s41467-021-25960-2)

```{r}
# Pseudobulk TBET expression
seurat$donor_id_annotation <- str_c(seurat$donor_id, seurat$annotation_20220619, sep = ";")
pseudobulk_tbx21 <- AverageExpression(
  seurat,
  features = "TBX21",
  group.by = "donor_id_annotation"
)$RNA
pseudobulk_tbx21 <- pseudobulk_tbx21 %>%
  t %>%
  as.data.frame %>%
  rownames_to_column(var = "donor_id_annotation") %>% # include age group as shape
  separate(col = donor_id_annotation, into = c("donor_id", "annotation"), sep = ";")
pseudobulk_tbx21_gg <- pseudobulk_tbx21 %>%
  ggplot(aes(fct_reorder(annotation, V1), V1, color = annotation)) +
    geom_boxplot(fill = NA, outlier.shape = NA) +
    geom_jitter(size = 0.25) +
    labs(y = "TBX21 expression") +
    scale_color_manual(values = cols_annotation, breaks = cols_annotation)
(pseudobulk_tbx21_gg <- theme_pseudobulk(pseudobulk_tbx21_gg) + coord_flip())
  


# Pseudobulk TBET activity
pseudobulk_tbet_activity_df <- seurat@meta.data %>%
  group_by(donor_id, annotation_20220619) %>%
  summarize(Tbet_activity = mean(TBET_activity))
pseudobulk_tbet_activity_gg <- pseudobulk_tbet_activity_df %>%
  ggplot(aes(fct_reorder(annotation_20220619, Tbet_activity), Tbet_activity, color = annotation_20220619)) +
    geom_boxplot(fill = NA, outlier.shape = NA) +
    geom_jitter(size = 0.25) +
    labs(y = "TBX21 activity (pySCENIC)") +
    scale_color_manual(values = cols_annotation, breaks = cols_annotation)
(pseudobulk_tbet_activity_gg <- theme_pseudobulk(pseudobulk_tbet_activity_gg) +
  coord_flip())
```


We can now visualize the correlation between TBET expression and activity:

```{r}
pseudobulk_tbx21$annotation_20220619 <- pseudobulk_tbx21$annotation
pseudobulk_tbx21_join <- left_join(
  pseudobulk_tbx21,
  pseudobulk_tbet_activity_df,
  by = c("donor_id", "annotation_20220619")
)
colnames(pseudobulk_tbx21_join)[colnames(pseudobulk_tbx21_join) == "V1"] <- "Tbet_expression"
(tbet_expression_activity_corr <- pseudobulk_tbx21_join %>%
  ggplot(aes(Tbet_expression, Tbet_activity)) +
    geom_smooth(method = "lm", se = FALSE, color = "gray25") +
    geom_point(aes(color = annotation_20220619)) +
    stat_correlation(mapping = use_label(c("R", "P")), size = 3.5) +
    scale_color_manual(values = cols_annotation, breaks = cols_annotation) +
    labs(x = "TBX21 expression", y = "TBX21 activity (pySCENIC)") + 
    theme_classic() +
    theme(axis.text = element_text(size = 7), axis.title = element_text(size = 8)))
```


Let us infer TBET activity using the list of genes that positively correlate with TBET:

```{r}
# Pseudobulk Tbet-dependent genes
pseudobulk_tbx21$annotation_20220619 <- pseudobulk_tbx21$annotation
seurat <- AddModuleScore_UCell(
  seurat,
  features = list(TBET_positive = pos_genes)
)
pseudobulk_tbet_pos_df <- seurat@meta.data %>%
  group_by(donor_id, annotation_20220619) %>%
  summarize(Tbet_activity_pos = mean(TBET_positive_UCell))
pseudobulk_tbx21_join2 <- left_join(
  pseudobulk_tbx21,
  pseudobulk_tbet_pos_df,
  by = c("donor_id", "annotation_20220619")
)
colnames(pseudobulk_tbx21_join2)[colnames(pseudobulk_tbx21_join2) == "V1"] <- "Tbet_expression"
pseudobulk_tbet_activity_pos_gg <- pseudobulk_tbet_pos_df %>%
  ggplot(aes(fct_reorder(annotation_20220619, Tbet_activity_pos), Tbet_activity_pos, color = annotation_20220619)) +
  geom_boxplot(fill = NA, outlier.shape = NA) +
  geom_jitter(size = 0.25) +
  labs(y = "Inducing TBX21 activity") +
  scale_color_manual(values = cols_annotation, breaks = cols_annotation) +
  coord_flip()
(pseudobulk_tbet_activity_pos_gg <- theme_pseudobulk(pseudobulk_tbet_activity_pos_gg) +
  coord_flip())


# Scatterplot
tbet_expression_activity_corr3 <- pseudobulk_tbx21_join2 %>%
  ggplot(aes(Tbet_expression, Tbet_activity_pos)) +
    geom_smooth(method = "lm", se = FALSE, color = "gray25") +
    geom_point(aes(color = annotation_20220619)) +
    stat_correlation(mapping = use_label(c("R", "P")), size = 3.5) +
    scale_color_manual(values = cols_annotation, breaks = cols_annotation) +
    labs(x = "TBX21 expression", y = "Inducing TBX21 activity") + 
    theme_classic() +
    theme(axis.text = element_text(size = 7), axis.title = element_text(size = 8))
tbet_expression_activity_corr3
```


# UMAP

Let us visualize TBET expression and activity in the UMAP. To do so, we will use [Nebulosa](https://academic.oup.com/bioinformatics/article/37/16/2485/6103785), which "uses a weighted kernel density estimation to recover signals lost through drop-out or low expression". This method is particularly suited to visualize the expression of transcription factors (TF), such as TBX21.

```{r}
# Rerun UMAP
seurat <- RunUMAP(seurat, reduction = "harmony", dims = 1:30)


# Plot UMAP annotation
umap_annotation <- DimPlot(seurat, pt.size = 0.01, raster = FALSE) +
  labs(x = "UMAP1", y = "UMAP2") +
  scale_color_manual(values = cols_annotation, breaks = names(cols_annotation)) +
  theme(axis.ticks = element_blank(),
        axis.text = element_blank(),
        legend.text = element_text(size = 8),
        axis.title = element_text(size = 8))
(umap_annotation <- ggrastr::rasterise(umap_annotation, dpi = 600))


# Plot UMAP TBET expression
t_bet_expression <- plot_density(
  seurat,
  "TBX21"
) +
  scale_color_viridis_c(option = "inferno") +
  labs(x = "UMAP1", y = "UMAP2", color = "TBET expression\n(density)")
(t_bet_expression <- theme_umap(t_bet_expression))
t_bet_expression <- ggrastr::rasterise(t_bet_expression, dpi = 600)


# Plot UMAP of TBET activity with Nebulosa
t_bet_activity <- plot_density(
  seurat,
  "TBET_activity"
) +
  scale_color_viridis_c(option = "viridis") +
  labs(x = "UMAP1", y = "UMAP2", color = "TBET activity\n(density)")
(t_bet_activity <- theme_umap(t_bet_activity))
t_bet_activity <- ggrastr::rasterise(t_bet_activity, dpi = 600)


# Regulons
targets_tbx21 <- regulons$gene[regulons$`TBX21(+)` == 1]
seurat$annotation_20220619 <- factor(
  seurat$annotation_20220619,
  levels = rev(names(cols_annotation))
)
Idents(seurat) <- "annotation_20220619"
dot_plot <- DotPlot(seurat, features = rev(targets_tbx21)) +
  scale_color_viridis_c(option = "inferno") +
  theme(
    axis.title = element_blank(),
    axis.text.y = element_text(size = 8),
    legend.text = element_text(size = 7),
    legend.title = element_text(size = 8),
    axis.text.x = element_text(size = 7, angle = 45, vjust = 1, hjust = 1))
dot_plot
```


# Dot plots

In a previous analysis, we used the AddModuleScore function from Seurat to infer the activity of each cell using a set of genes that are predicted to be regulated by TBX21. This analysis suggests that TBX21 may be more active in FCRL4+ MBC. However, it is important to assess how the expression of the gene set is distributed across the different NBC/MBC populations, which will tell us which are the genes that are dominating the signal. This is particularly important in scRNA-seq data, because some genes in the signature might not be captured and thus will not contribute to the scoring.

```{r fig.wide=TRUE}
dot_plots <- map(list(pos_genes, neg_genes), \(x) {
  p <- Clustered_DotPlot(
    seurat,
    features = x,
    colors_use_exp = viridis::inferno(20),
    colors_use_idents = cols_annotation,
    row_label_size = 6,
    column_label_size = 5.5,
    legend_label_size = 6,
    legend_title_size = 6,
    plot_km_elbow = FALSE,
    flip = TRUE
  )
  p
})
```


# Save

```{r eval=FALSE}
dir.create(here("results/plots"), recursive = TRUE)
ggsave(
  filename = here("results/plots/TBET_expression_NBC_MBC_tonsil_boxplot.pdf"),
  plot = pseudobulk_tbx21_gg,
  width = 12,
  height = 9,
  units = "cm"
)
ggsave(
  filename = here("results/plots/TBET_activity_NBC_MBC_tonsil_boxplot.pdf"),
  plot = pseudobulk_tbet_activity_gg,
  width = 12,
  height = 9,
  units = "cm"
)
ggsave(
  filename = here("results/plots/TBET_activityTbet_dependent_positive_NBC_MBC_tonsil_boxplot.pdf"),
  plot = pseudobulk_tbet_activity_pos_gg,
  width = 12,
  height = 9,
  units = "cm"
)
ggsave(
  filename = here("results/plots/TBET_activity_expression_scatterplot_NBC_MBC_tonsil.pdf"),
  plot = tbet_expression_activity_corr3,
  width = 12,
  height = 9,
  units = "cm"
)
ggsave(
  filename = here("results/plots/annotation_NBC_MBC_tonsil.pdf"),
  plot = umap_annotation,
  width = 12,
  height = 9,
  units = "cm"
)
ggsave(
  filename = here("results/plots/TBET_activity_NBC_MBC_tonsil.pdf"),
  plot = t_bet_activity,
  width = 12,
  height = 9,
  units = "cm"
)
ggsave(
  filename = here("results/plots/TBET_expression_NBC_MBC_tonsil.pdf"),
  plot = t_bet_expression,
  width = 12,
  height = 9,
  units = "cm"
)

in2mm <- 25.4
path_save_dotplot_pos <- here("results/plots/dot_plot_TBET_dependent_genes_Philipp_NBC_MBC_tonsil_positive.pdf")
pdf(path_save_dotplot_pos, width = (260 / in2mm), height = (75 / in2mm), paper = "special")
print(dot_plots[[1]])
dev.off()

path_save_dotplot_neg <- here("results/plots/dot_plot_TBET_dependent_genes_Philipp_NBC_MBC_tonsil_negative.pdf")
pdf(path_save_dotplot_neg, width = (260 / in2mm), height = (75 / in2mm), paper = "special")
print(dot_plots[[2]])
dev.off()
```


# Session Information

```{r}
sessionInfo()
```

